### DISCUSSION TITLE
`token` value is `null` when using `getToken`

### DISCUSSION THREAD

**Discussion by deadcoder0904** [nextauthjs/next-auth#2964] [Help] [answered]
I'm still following https://spacejelly.dev/posts/how-to-make-twitter-api-requests-with-nextauth-js-session-tokens/ & I have `/api/twitter/search` endpoint that looks like:

```ts
import { NextApiRequest, NextApiResponse } from 'next'
import { getSession } from 'next-auth/react'
import { getToken } from 'next-auth/jwt'

export default async (req: NextApiRequest, res: NextApiResponse) => {
  try {
    const session = await getSession({ req })
    const token = await getToken({
      req,
      token: process.env.NEXTAUTH_SECRET,
    })

    console.log({ session })
    console.log({ token })
    return res.status(200).json({
      status: 'Ok',
      data: [],
    })
  } catch (e: unknown) {
    return res.status(400).json({
      status: (e as Error).message,
    })
  }
}
```

I get the `token` value to be `null` but I do have `session` value. How do I get the `token` value? The blog post used v3 so not sure if something changed in v4.

Full repro is on branch `next-auth-with-twitter-api-v2` → https://github.com/deadcoder0904/twitter-api-v2-3-legged-login-using-next-connect/tree/next-auth-with-twitter-api-v2

**Comment by balazsorban44** [comment, marked as answer]
You have a typo:
```js
getToken({
  req,
  secret: process.env.NEXTAUTH_SECRET
})
```

tip, you probably don't need both getSession and getToken. session is the subset of token in most cases, depending on what you do in your session callback

**Reply by deadcoder0904** [reply]
I'm so dumb ah. Typo hurts.

Bdw, `accessToken` & `refreshToken` are `undefined` when I log `token`. Is this because of [this](https://github.com/nextauthjs/next-auth/discussions/2886#discussioncomment-1423068)? Because I also `console.log(account)` & it doesn't have `account.access_token` & `account.refresh_token` at all in them. It only has:

```ts
{
  account: {
    provider: 'twitter',
    type: 'oauth',
    providerAccountId: 'basbasb',
    oauth_token: 'basbasb-123891283918923891',
    oauth_token_secret: '218989321923989'
  }
}
```

> tip, you probably don't need both getSession and getToken. session is the subset of token in most cases, depending on what you do in your session callback

they don't have any overlaps right now. idk the data i want to store yet so i'll probably refactor it later :)

**Reply by balazsorban44** [reply]
Twitter doesn't operate with OAuth 2 tokens, since it's using thebOAuth 1 specification.

**Reply by deadcoder0904** [reply]
@balazsorban44 so how come it works in the tutorial? it logs `accessToken` & `refreshToken` on there.

**Reply by balazsorban44** [reply]
I don't follow tutorials, but I implemented it, so I'm sure I am right. 😅 We basically don't map these fields, but give them to the user as-is. That's how Twitter returns them.

At the end it is just naming though. When Twitter finally makes OAuth 2 a thing on their side, this will become irrelevant. I already started playing with their beta.

**Reply by deadcoder0904** [reply]
@balazsorban44 you are not getting what I'm saying 😂

i'm saying that in the tutorial linked above the value of `accessToken` & `refreshToken` is not `undefined` but when I'm doing that exact same thing i get `undefined`. the only difference is i'm using next-auth v4.

the code looks like:

```ts
callbacks: {
    async jwt({ token, account }) {
      if (account) {
        console.log({ account })
        token[account.provider] = {
          accessToken: account.access_token,
          refreshToken: account.refresh_token,
        }
        console.log({ token })
      }

      return token
    },
  },
```

this means that `account.access_token` & `account.refresh_token` is `undefined` as you can  see `account` only has the following properties:

```ts
{
  account: {
    provider: 'twitter',
    type: 'oauth',
    providerAccountId: 'basbasb',
    oauth_token: 'basbasb-123891283918923891',
    oauth_token_secret: '218989321923989'
  }
}
```

is there something i'm doing wrong here?

**Reply by balazsorban44** [reply]
Yes. As said, Twitter is an OAuth 1 provider, and so it returns those tokens with different names. Please read above.

Here is a link to Twitter terminology https://developer.twitter.com/en/docs/authentication/oauth-1-0a/obtaining-user-access-tokens

**Reply by deadcoder0904** [reply]
yes i saw that & found this:

<img width="748" alt="Screenshot 2021-10-15 at 10 16 57 PM" src="https://user-images.githubusercontent.com/16436270/137523798-8f7dd280-8172-42c6-b62c-f26512c38c6f.png">

but not sure where do i get `refresh_token`? also, `access_token` is `oauth_token`?

**Reply by deadcoder0904** [reply]
besides since this library is typed, i don't get `oauth_token` as a method when i do `account.`

i only get `account.access_token` & `account.refresh_token` when i do autocompletion.

**Reply by balazsorban44** [reply]
I'm actually not familiar with the OAuth 1 flow other than that difference. I haven't used the Twitter API  in the past. You might have to do your own research how to use the returned tokens.

**Reply by balazsorban44** [reply]
For type augmentation, refer to https://next-auth.js.org/getting-started/typescript
